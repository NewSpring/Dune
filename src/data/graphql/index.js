// @flow
import { graphql as graphqljs } from "graphql";
import { makeExecutableSchema } from "graphql-tools";
import { print } from "graphql/language/printer";
import type { ExecutionResult, DocumentNode } from "graphql";

export const typeDefs = `
type Response {
  code: Int!
  message: String
}

type Page {
  id: Int!
  layout: String!
  blocks: [Block]
}

type Block {
  id: Int!
  path: String!
  zone: String!,
  order: Int!,
}

# the schema allows the following query:
type Query {
  sample: Response
  getRouteInfo(path: String!): Page
}

schema {
  query: Query
}
`;

type ISample = {
  code: number,
  message: string,
};

type IBlock = {
  id: number,
  path: string,
  order: number,
  zone: string,
};

type IPage = {
  id: number,
  layout: string,
  blocks: IBlock[],
};

export const resolvers = {
  Query: {
    sample: (): ISample => ({ code: 200, message: "hello world" }),
    getRouteInfo: () => {
      return {
        id: 1,
        layout: "Horizontal",
        blocks: [
          { path: "HelloWorld", id: 3, zone: "main", order: 1 },
          { path: "HelloWorld", id: 2, zone: "secondary", order: 1 },
          { path: "Counter", id: 1, zone: "secondary", order: 0 },
        ],
      };
    },
  },
  Response: {
    code: ({ code }: ISample): number => code,
    message: ({ message }: ISample): string => message,
  },
  Page: {
    id: ({ id }: IPage) => id,
    layout: ({ layout }: IPage) => layout,
    blocks: ({ blocks }: IPage) => blocks,
  },
  Block: {
    id: ({ id }: IBlock) => id,
    zone: ({ zone }: IBlock) => zone,
    order: ({ order }: IBlock) => order,
    path: ({ path }: IBlock) => path,
  },
};

export const schema = makeExecutableSchema({ typeDefs, resolvers });

export const safeQuery = (query: string | DocumentNode): string =>
  typeof query === "string" ? query : print(query);
// this is a curried schema execution function
// use it to execute queries for SSR and testing
export const graphql = (
  query: string | DocumentNode,
  root?: mixed,
  context?: mixed,
  variables?: ?{ [key: string]: mixed },
  operationName?: ?string,
): Promise<ExecutionResult> =>
  graphqljs(schema, safeQuery(query), root, context, variables, operationName);
