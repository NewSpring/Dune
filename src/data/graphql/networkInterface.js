//@flow
import { createBatchingNetworkInterface } from "react-apollo";

export default () =>
  createBatchingNetworkInterface({
    uri: "http://localhost:3000/graphql",
    batchInterval: 10,
  });
