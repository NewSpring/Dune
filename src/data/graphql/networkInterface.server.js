// @flow
/*
 * NOTE
 * 
 * middlewares are not pure in apollo client
 * each ware mutates the request (or the result)
 * and calls next() indicating the next ware can run
 *
 * it isn't worth our time currently to refactor that to make it
 * functional (it could be a fun project though)
 *
 * apollo-client has an odd API for middleware / afterware that comes
 * from ad-hoc dev of the NetworkInterface feature set. Cleaning it up
 * could be a great PR (it would be a breaking change so backwards compat
 * would be critical)
 *
 */
import { slice } from "ramda";
import type { DocumentNode, ExecutionResult } from "graphql";
import { graphql } from "./";

// shape of request from apollo client
export type IRequest = {
  query: DocumentNode,
  variables?: { [key: string]: any },
  operationName?: string,
};

export type IApplyRequest = {
  request?: IRequest,
  response?: ExecutionResult,
};

export type IMiddlewareRequest = {|
  request: IRequest,
  context: mixed,
|};

export type IMiddleware = {
  applyMiddleware(request: IMiddlewareRequest, next: Function): void,
};

export type IAfterwareResponse = {
  response: ExecutionResult,
};

export type IAfterware = {
  applyAfterware(response: IAfterwareResponse, next: Function): void,
};

export type IWare = IMiddleware | IAfterware;
export type IPayload = IMiddlewareRequest | IAfterwareResponse;

export interface INetworkInterface {
  _opts: mixed,
  _middlewares: IMiddleware[],
  _afterwares: IAfterware[],

  use: (middlewares: IMiddleware[]) => INetworkInterface,
  useAfter: (middlewares: IAfterware[]) => INetworkInterface,

  query: (request: IRequest) => Promise<ExecutionResult>,
}

export function NetworkInterface(options: mixed = {}) {
  if (!(this instanceof NetworkInterface)) return new NetworkInterface(options);
  this._middlewares = [];
  this._afterwares = [];
  this._opts = options;
  return this;
}

export const apply = (middlewares: IWare[], payload: IPayload): Promise<> =>
  new Promise(resolve => {
    const funcs = slice(0, middlewares.length, middlewares);
    const next = () => {
      if (funcs.length > 0) {
        const f = funcs.shift();
        if (f && f.applyMiddleware && payload.request)
          f.applyMiddleware(payload, next);
        if (f && f.applyAfterware && payload.response)
          f.applyAfterware(payload, next);
      } else {
        resolve(payload);
      }
    };
    next();
  });

NetworkInterface.prototype.query = function(
  request: IRequest,
): Promise<ExecutionResult> {
  // XXX context handling
  return apply(this._middlewares, { request })
    .then(
      ({
        request: { query, variables, operationName },
        context,
      }: IMiddlewareRequest) =>
        graphql(query, {}, context, variables, operationName),
    )
    .then(response => apply(this._afterwares, { response }))
    .then(({ response }: IAfterwareResponse) => response);
};

NetworkInterface.prototype.use = function(
  middlewares: IMiddleware[],
): INetworkInterface {
  this._middlewares = this._middlewares.concat(
    middlewares.filter(
      (x: IMiddleware) => typeof x.applyMiddleware === "function",
    ),
  );
  return this;
};

NetworkInterface.prototype.useAfter = function(
  middlewares: IAfterware[],
): INetworkInterface {
  this._afterwares = this._afterwares.concat(
    middlewares.filter(
      (x: IAfterware) => typeof x.applyAfterware === "function",
    ),
  );

  return this;
};

export default (options: mixed): INetworkInterface => NetworkInterface(options);
