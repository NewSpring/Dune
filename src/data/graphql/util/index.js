// @flow

export const platform = () => "web";
export const bundleId = () =>
  window && window.location && window.location.hostname
    ? window.location.hostname
    : "";
export const version = () =>
  process && process.env && process.env.BUILD ? process.env.BUILD : "";
