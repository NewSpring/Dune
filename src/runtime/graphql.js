// @flow
import { graphqlLambda, graphiqlLambda } from "graphql-server-lambda";

import { schema } from "../data/graphql";

// XXX move these to a typings file for lambda
import type { ILambdaEvent, ILambdaContext } from "./ssr/withApp";

export type ILambdaOutput = {
  headers: { [key: string]: string },
};

export type ILambdaCallback = (error: mixed, output: ILambdaOutput) => void;

export type IContext = {
  userId?: string,
  ip?: string,
  isDev: boolean,
};

export const createContextFromRequestEvent = (
  event: ILambdaEvent,
): IContext => {
  const { headers, requestContext } = event;
  return {
    userId: headers && headers.authorization,
    ip: requestContext &&
      requestContext.identity &&
      requestContext.identity.sourceIp,
    isDev: requestContext && requestContext.stage === "dev",
  };
};

// enable cors
export const withCors = (cb: ILambdaCallback): ILambdaCallback => (
  error,
  output,
) => {
  output.headers["Access-Control-Allow-Origin"] = "*";
  cb(error, output);
};

export const graphqlEndpoint = (
  event: ILambdaEvent,
  ctx: ILambdaContext,
  cb: ILambdaCallback,
) => graphqlLambda({ schema })(event, ctx, withCors(cb));

export const graphiqlEndpoint = (
  event: ILambdaEvent,
  ctx: ILambdaContext,
  cb: ILambdaCallback,
) => {
  const stage = event.isOffline ? "" : `/${event.requestContext.stage}`;
  return graphiqlLambda({ endpointURL: `${stage}/graphql` })(event, ctx, cb);
};
