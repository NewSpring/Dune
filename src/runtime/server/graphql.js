// @flow
import { graphql as graphqljs } from "graphql";
import { makeExecutableSchema } from "graphql-tools";
import { graphqlLambda, graphiqlLambda } from "graphql-server-lambda";

import type { ExecutionResult } from "graphql";
// XXX move these to a typings file for lambda
import type { ILambdaEvent, ILambdaContext } from "./ssr/withApp";

export const typeDefs = `
type Response {
  code: Int!
  message: String
}

# the schema allows the following query:
type Query {
  sample: Response
}

schema {
  query: Query
}
`;

type ISample = {
  code: number,
  message: string,
};

export const resolvers = {
  Query: {
    sample: (): ISample => ({ code: 200, message: "hello world" }),
  },
  Response: {
    code: ({ code }: ISample): number => code,
    message: ({ message }: ISample): string => message,
  },
};

export const schema = makeExecutableSchema({ typeDefs, resolvers });

// this is a curried schema execution function
// use it to execute queries for SSR and testing
export const graphql = (
  query: string,
  root?: mixed,
  context?: mixed,
  variables?: ?{ [key: string]: mixed },
  operationName?: ?string,
): Promise<ExecutionResult> =>
  graphqljs(schema, query, root, context, variables, operationName);

export type ILambdaOutput = {
  headers: { [key: string]: string },
};
export type ILambdaCallback = (error: mixed, output: ILambdaOutput) => void;

// enable cors
export const withCors = (cb: ILambdaCallback): ILambdaCallback => (
  error,
  output,
) => {
  output.headers["Access-Control-Allow-Origin"] = "*";
  cb(error, output);
};

export const graphqlEndpoint = (
  event: ILambdaEvent,
  ctx: ILambdaContext,
  cb: ILambdaCallback,
) => graphqlLambda({ schema })(event, ctx, withCors(cb));

export const graphiqlEndpoint = (
  event: ILambdaEvent,
  ctx: ILambdaContext,
  cb: ILambdaCallback,
) => {
  const stage = event.isOffline ? "" : `/${event.requestContext.stage}`;
  return graphiqlLambda({ endpointURL: `${stage}/graphql` })(event, ctx, cb);
};
