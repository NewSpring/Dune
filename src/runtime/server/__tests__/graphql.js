import { typeDefs, resolvers, withCors, graphql } from "../graphql";

describe("typeDefs", () => {
  it("should match a snapshot", () => {
    expect(typeDefs).toMatchSnapshot();
  });
});

describe("resolvers", () => {
  const { Query, Response } = resolvers;

  describe("Query", () => {
    const { sample } = Query;
    it("has an expected shape", () => {
      expect(sample()).toEqual({ code: 200, message: "hello world" });
    });
  });

  describe("Response", () => {
    const data = Query.sample();
    const { code, message } = Response;
    it("has a code method", () => {
      expect(code(data)).toEqual(200);
    });
    it("has a message method", () => {
      expect(message(data)).toEqual("hello world");
    });
  });
});

describe("graphql", () => {
  it("has a sample root level query field", async () => {
    const { data } = await graphql("{ sample { code message } }");
    expect(data).toEqual({ sample: { code: 200, message: "hello world" } });
  });
});

describe("withCors", () => {
  it("sets the ACAO on the output headers", () => {
    const cb = jest.fn(() => {});
    const output = { headers: {} };
    const error = "foo";

    withCors(cb)(error, output);
    expect(cb).toBeCalledWith("foo", {
      headers: {
        "Access-Control-Allow-Origin": "*",
      },
    });
  });
});
