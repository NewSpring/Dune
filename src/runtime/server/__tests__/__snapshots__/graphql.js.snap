// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`typeDefs should match a snapshot 1`] = `
"
type Response {
  code: Int!
  message: String
}

# the schema allows the following query:
type Query {
  sample: Response
}

schema {
  query: Query
}
"
`;
